stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -2535,10 +3132,92 @@ impl NakamotoChainState {
         let new_block_hash = new_tip.block_hash();
         let index_block_hash = new_tip.block_id();
 
+        let mut marf_keys = vec![];
+        let mut marf_values = vec![];
+
+        if new_tenure {
+            // make the coinbase height point to this tenure-start block
+            marf_keys.push(nakamoto_keys::ongoing_tenure_coinbase_height(
+                coinbase_height,
+            ));
+            marf_values.push(nakamoto_keys::make_block_id_value(&new_tip.block_id()));
+
+            // point this tenure to its start block
+            marf_keys.push(nakamoto_keys::tenure_start_block_id(
+                &new_tip.consensus_hash,
+            ));
+            marf_values.push(nakamoto_keys::make_block_id_value(&new_tip.block_id()));
+
+            // record coinbase height of this tenure
+            marf_keys.push(nakamoto_keys::coinbase_height(&new_tip.consensus_hash));
+            marf_values.push(nakamoto_keys::make_u64_value(coinbase_height));
+
+            // record that this previous tenure is done
+            let Some(tenure_change_tx) = new_block.get_tenure_change_tx_payload() else {
+                // should be unreachable
+                error!(
+                    "Block {} is a tenure-change block, but does not contain a tenure-change tx",
+                    new_tip.block_id()
+                );
+                return Err(ChainstateError::InvalidStacksBlock(
+                    "Tenure-change block does not have a tenure-change tx".into(),
+                ));
+            };
+
+            marf_keys.push(nakamoto_keys::finished_tenure_consensus_hash(
+                &tenure_change_tx.prev_tenure_consensus_hash,
+            ));
+            marf_values.push(nakamoto_keys::make_bool_value(true));
+
+            // record parent tenure linkage
+            marf_keys.push(nakamoto_keys::parent_tenure_consensus_hash(
+                &tenure_change_tx.tenure_consensus_hash,
+            ));
+            marf_values.push(nakamoto_keys::make_consensus_hash_value(
+                &tenure_change_tx.prev_tenure_consensus_hash,
+            ));
+
+            // record last block-found tenure
+            let block_found_tenure_id = NakamotoTenureId {
+                burn_view_consensus_hash: tenure_change_tx.burn_view_consensus_hash.clone(),
+                block_id: new_tip.block_id(),
+            };
+
+            marf_keys.push(nakamoto_keys::block_found_tenure_id(
+                &tenure_change_tx.tenure_consensus_hash,
+            ));
+            marf_values.push(nakamoto_keys::make_tenure_id_value(&block_found_tenure_id));
+        }
+
+        if let Some(tenure_tx) = new_block.get_tenure_tx_payload() {
+            // either a block-found or a tenure-extend, but we have a new tenure ID in this fork
+            let tenure_id = NakamotoTenureId {
+                burn_view_consensus_hash: tenure_tx.burn_view_consensus_hash.clone(),
+                block_id: new_tip.block_id(),
+            };
```

Comment:
This is confusing to me -- my understanding was that tenure extends _do not_ define a new tenure. The tenure of a given block would always be the consensus hash that elected the miner.

Reply:
You're correct -- tenure-extends do not qualify as new tenures.  However, per the docs, the rows of the `nakamoto_tenures` table (and corresponding `NakamotoTenure` struct) encode a tenure-change event in the chain history, be it a `BlockFound` or `Extended` tenure-change.  The `NakamotoTenureId` is an identifier for a row in this table.

Perhaps some renaming is in order?
* `nakamoto_tenures` table --> `nakamoto_tenure_events` table
* `NakamotoTenure` struct --> `NakamotoTenureEvent`
* `NakamotoTenureId` struct --> `NakamotoTenureEventId`